pkg/log/logger.go

```diff
@@ -0,0 +1,222 @@
+	WithContext(ctx context.Context) Logger
+	WithFields(Fields) Logger
+}
+
+type GosoLogger interface {
+	Logger
+	Option(opt ...Option) error
+}
+
+type LoggerSettings struct {
+	Handlers []HandlerSettings `cfg:"handlers"`
+}
+
+type HandlerSettings struct {
+	Type string `cfg:"type"`
+}
+
+type gosoLogger struct {
+	clock        clock.Clock
+	data         Data
+	ctxResolvers []ContextFieldsResolver
+	handlers     []Handler
+}
+
+func NewLogger() *gosoLogger {
+	return NewLoggerWithInterfaces(clock.NewRealClock(), []Handler{})
+}
+
+func NewLoggerWithInterfaces(clock clock.Clock, handlers []Handler) *gosoLogger {
+	return &gosoLogger{
+		clock: clock,
+		data: Data{
+			Context:       nil,
+			Channel:       "main",
+			ContextFields: make(map[string]interface{}),
+			Fields:        make(map[string]interface{}),
+		},
+		ctxResolvers: nil,
+		handlers:     handlers,
+	}
+}
+
+func (l *gosoLogger) Option(options ...Option) error {
+	for _, opt := range options {
+		if err := opt(l); err != nil {
+			return fmt.Errorf("can not apply option %T", err)
+		}
+	}
+
+	return nil
+}
+
+func (l *gosoLogger) Debug(format string, args ...interface{}) {
+	l.log(PriorityDebug, format, args, nil)
+}
+
+func (l *gosoLogger) Info(format string, args ...interface{}) {
+	l.log(PriorityInfo, format, args, nil)
+}
+
+func (l *gosoLogger) Warn(format string, args ...interface{}) {
+	l.log(PriorityWarn, format, args, nil)
+}
+
+func (l *gosoLogger) Error(format string, args ...interface{}) {
+	err := fmt.Errorf(format, args...)
+	l.log(PriorityError, format, args, err)
+}
+
+func (l *gosoLogger) WithChannel(channel string) Logger {
+	cpy := l.copy()
+	cpy.data.Channel = channel
+
+	return cpy
+}
+
+func (l *gosoLogger) WithContext(ctx context.Context) Logger {
+	if ctx == nil {
+		return l
+	}
+
+	cpy := l.copy()
+	cpy.data.Context = ctx
+
+	for _, r := range l.ctxResolvers {
+		newContextFields := r(ctx)
+		cpy.data.ContextFields = mergeFields(cpy.data.ContextFields, newContextFields)
+	}
+
+	return cpy
+}
+
+func (l *gosoLogger) WithFields(fields Fields) Logger {
+	cpy := l.copy()
+	cpy.data.Fields = mergeFields(l.data.Fields, fields)
+
+	return cpy
+}
+
+func (l *gosoLogger) copy() *gosoLogger {
+	return &gosoLogger{
+		clock:        l.clock,
+		data:         l.data,
+		ctxResolvers: l.ctxResolvers,
+		handlers:     l.handlers,
+	}
+}
+
+func (l *gosoLogger) log(level int, msg string, args []interface{}, err error) {
+	timestamp := l.clock.Now()
+
+	for _, handler := range l.handlers {
+		if handler.Level() > level {
+			continue
+		}
+
+		handlerChannels := handler.Channels()
+
+		if !l.shouldLogToChannels(l.data.Channel, handlerChannels) {
+			continue
+		}
+
+		if err := handler.Log(timestamp, level, msg, args, err, l.data); err != nil {
+			l.err(err)
+		}
+	}
+}
+
+func (l *gosoLogger) shouldLogToChannels(current string, channels []string) bool {
+	if len(channels) == 0 {
+		return true
+	}
+
+	for _, channel := range channels {
```

Comment:
Just an idea: would it be better to read and understand the different cases when using `funk.ContainsString()` with either `*`, `channel` or `!channel`?

Reply:
would be an alternative, yes